src/java.base/share/classes/java/lang/foreign/SegmentAllocator.java

```diff
@@ -257,89 +257,162 @@ default MemorySegment allocateFrom(AddressLayout layout, MemorySegment value) {
+     * @param sourceOffset the starting offset, in bytes, of the source segment.
+     * @param srcElementCount the number of elements in the source segment to be copied.
+     * @return a segment for the newly allocated memory block.
+     * @throws IllegalArgumentException if the element layouts have different sizes, if the source segment/offset are
+     * <a href="MemorySegment.html#segment-alignment">incompatible with the alignment constraint</a> in the source
+     * element layout, or if the source element layout alignment is greater than its size.
+     * @throws IllegalStateException if the {@linkplain MemorySegment#scope() scope} associated with {@code source} is not
+     * {@linkplain MemorySegment.Scope#isAlive() alive}.
+     * @throws WrongThreadException if this method is called from a thread {@code T},
+     * such that {@code source.isAccessibleBy(T) == false}.
+     * @throws IndexOutOfBoundsException if {@code elementCount * sourceElementLayout.byteSize()} overflows.
+     * @throws IndexOutOfBoundsException if {@code sourceOffset > source.byteSize() - (elementCount * sourceElementLayout.byteSize())}.
+     * @throws IndexOutOfBoundsException if either {@code sourceOffset} or {@code elementCount} are {@code < 0}.
+     */
+    @ForceInline
+    default MemorySegment allocateFrom(ValueLayout elementLayout, MemorySegment source,
+                                       ValueLayout sourceElementLayout, long sourceOffset, long srcElementCount) {
+        Objects.requireNonNull(source);
+        Objects.requireNonNull(sourceElementLayout);
+        Objects.requireNonNull(elementLayout);
+        MemorySegment dest = allocateNoInit(elementLayout, srcElementCount);
+        MemorySegment.copy(source, sourceElementLayout, sourceOffset, dest, elementLayout, 0, srcElementCount);
+        return dest;
+    }
+
     /**
      * Allocates a memory segment with the given layout and initializes it with the given byte elements.
-     * @implSpec the default implementation for this method calls {@code this.allocate(layout, array.length)}.
+     * @implSpec the default implementation for this method is equivalent to the following code:
+     * {@snippet lang = java:
+     * this.allocateFrom(layout, MemorySegment.ofArray(array),
+     *                   layout.withOrder(ByteOrder.nativeOrder()),
+     *                   0, array.length)
+     *}
      * @param elementLayout the element layout of the array to be allocated.
      * @param elements the byte elements to be copied to the newly allocated memory block.
      * @return a segment for the newly allocated memory block.
      */
+    @ForceInline
     default MemorySegment allocateFrom(ValueLayout.OfByte elementLayout, byte... elements) {
-        return copyArrayWithSwapIfNeeded(elements, elementLayout);
+        return allocateFrom(elementLayout, MemorySegment.ofArray(elements),
+                elementLayout.withOrder(ByteOrder.nativeOrder()), 0, elements.length);
```

Comment:
Can't the source layout here just be `JAVA_BYTE`? Since we're copying from a java byte array. (similar for the other overloads).

```suggestion
                JAVA_BYTE, 0, elements.length);
```

Reply:
> No, if we do not propagate the original layout, some of the alignment checks are going to be missing.

I don't get this. I'm not saying we don't propagate the original layout, I'm just saying that the source layout should be `JAVA_BYTE` here.

The alignment checks of the destination layout only apply to the destination segment, right? Or are you saying that if the destination layout is hyper-aligned (> 1 byte), we should fail because the byte[] elements are only aligned to 1 byte?

Reply:
> > No, if we do not propagate the original layout, some of the alignment checks are going to be missing.
> 
> I don't get this. I'm not saying we don't propagate the original layout, I'm just saying that the source layout should be `JAVA_BYTE` here.
> 
> The alignment checks of the destination layout only apply to the destination segment, right? Or are you saying that if the destination layout is hyper-aligned (> 1 byte), we should fail because the byte[] elements are only guaranteed to be aligned to 1 byte?

I believe I just misinterpreted what you were saying - as I made essentially the same point below :-)